src/Protocols/ReqResp.hs

```diff
@@ -0,0 +1,46 @@
+{- |
+Simple protocol for request-response communication.
+The forward channel channel has type @Signal dom (Maybe req)@ and is used to send requests.
+The backward channel has type @Signal dom (Maybe resp)@ and is used to send responses.
+The protocol must obey the following rules:
+* When the forward channel is @Just a@, it must not change until the transaction is completed.
```

Comment:
This implies that this protocol can not be pipelined. We can relax this constraint and do two versions of ReqResp (pipelined and not pipelined). Or we can use DSignal to enforce different pipeline depths on the type level.

Reply:
Why couldn't it be pipelined? I think you can write a skid buffer that handles this. Essentially store forward in a register. Send contents of the register to the sink. When the sink responds either send it directly to the source and clear the register or put the response in a register as well.

Reply:
Perhaps we should then just have
```haskell
data ReqResp (dom :: C.Domain) (req :: Type) (resp :: Type) = (Df dom req, Reverse (Df dom resp))
```
a kind of Axi4LiteLite

Reply:
Enforcing pipeline depth on the type level seems to be more useful than just a blanket declaration "you can send stuff, which we call a request, and you can receive stuff, which we call a response", which seems to be what Christiaans `ReqRespPipelined` is.

Avalon Stream does pre-negotiated pipelining. You can configure that when you drop _ready_, you can still receive data in the next _n_ cycles, so the pipeline of the sender can empty. Constructions like that. Maybe it can serve as inspiration.

Alternatively, we can tackle pipelining in a different PR.

Reply:
I don't think separating the channels is a good idea as it directly contradicts the purpose of `circuit-notation` (Combining forward and backward channels in a single binder).

How about separating request acknowledgement and response timing?:
```hs
data ReqResp (dom :: C.Domain) (req :: Type) (resp :: Type)

instance Protocol (ReqResp dom req resp) where
  -- | Forward channel for ReqResp protocol
  type Fwd (ReqResp dom req resp) = C.Signal dom (Maybe req)

  -- | Backward channel for ReqResp protocol
  type Bwd (ReqResp dom req resp) = C.Signal dom (Bool, Maybe resp)
  ```
Here the `Bool` would indicate a request acknowledgement and `Maybe resp` could be `Just resp` later.
Combinatorial circuits could then return `(isJust resp, resp)`.

Reply:
How would you still be able to tell which response corresponds to which request?

Personally, I think a "simple" protocol that doesn't allow pipelining at the protocol level¹ and a separate more complex protocol that tackles pipelining in a neat and efficient way is a perfectly fine solution. This PR can be the simple variant.

¹ As Rowan indicates above, you can still accomodate pipelining with constructions like skid buffers, but the protocol would be oblivious to this and it would in general be less efficient than pipelining being part of the protocol.